api/src/auth/endpoint_access_util.py

```diff
@@ -0,0 +1,104 @@
+import logging
+from typing import Any
+
+from src.constants.lookup_constants import Privilege
+from src.db.models.agency_models import Agency
+from src.db.models.competition_models import Application
+from src.db.models.entity_models import Organization
+from src.db.models.user_models import Role, User
+
+logger = logging.getLogger(__name__)
+
+
+def get_roles_for_agency(user: User, agency: Agency) -> list[Role]:
+    for user_agency in user.user_agencies:
+        if user_agency.agency_id == agency.agency_id:
+            return user_agency.roles
+
+    return []
+
+
+def get_roles_for_org(user: User, organization: Organization) -> list[Role]:
+    for user_organization in user.organizations:
+        if user_organization.organization_id == organization.organization_id:
+            return user_organization.roles
+
+    return []
+
+
+def get_roles_for_app(user: User, application: Application) -> list[Role]:
+    roles = []
+    # If the app has an org, add any org roles for the user
+    if application.organization is not None:
+        roles.extend(get_roles_for_org(user, application.organization))
+
+    for application_user in application.application_users:
+        if application_user.user_id == user.user_id:
+            roles.extend(application_user.roles)
+
+    return roles
+
+
+def get_roles_for_resource(
+    user: User, resource: Organization | Application | Agency | None
+) -> list[Role]:
+    roles = []
+
+    # Add any internal roles
+    roles.extend(user.internal_roles)
+
+    # As we add types of resources, just need to add here
+    if isinstance(resource, Organization):
+        roles.extend(get_roles_for_org(user, resource))
+    elif isinstance(resource, Application):
+        roles.extend(get_roles_for_app(user, resource))
+    elif isinstance(resource, Agency):
+        roles.extend(get_roles_for_agency(user, resource))
+
+    return roles
+
+
+def can_access(
```

Comment:
An idea for the following tickets where we'll add usage of this function, we might want a function that wraps this like:

```py
def validate_privileges(...same params):
     if not can_access(...):
          raise_flask_error(403, "Access Denied - Missing required privilege") # maybe a validation error added.
```
That way we don't need to copy-paste that `raise_flask_error` to a bunch of places. Not needed yet, but an idea for the next tickets.